src/middleware/route-ratelimit.ts

```diff
@@ -1,48 +1,105 @@
+/*
+  This file controls the request-per-minute (RPM) rate limits.
+
+  It is assumed that this middleware is run AFTER the auth.js middleware which
+  checks for Basic auth. If the user adds the correct Basic auth to the header
+  of their API request, they will get pro-tier rate limits. By default, the
+  freemium rate limits apply.
+*/
+
 import * as express from "express"
 const RateLimit = require("express-rate-limit")
 
+// Used for debugging.
+const util = require("util")
+util.inspect.defaultOptions = { depth: 3 }
+
 // Set max requests per minute
-const maxRequests = process.env.RATE_LIMIT_MAX_REQUESTS ? parseInt(process.env.RATE_LIMIT_MAX_REQUESTS) : 60
+const maxRequests = process.env.RATE_LIMIT_MAX_REQUESTS
+  ? parseInt(process.env.RATE_LIMIT_MAX_REQUESTS)
+  : 60
+
+// Pro-tier rate limits are 10x the freemium limits.
+const PRO_RPM = 10 * maxRequests
 
 // Unique route mapped to its rate limit
 const uniqueRateLimits: any = {}
 
+interface Request extends express.Request {
+  locals: any
+}
+
 const routeRateLimit = function(
-  req: express.Request,
+  req: Request,
   res: express.Response,
   next: express.NextFunction
 ) {
   // Disable rate limiting if 0 passed from RATE_LIMIT_MAX_REQUESTS
   if (maxRequests === 0) return next()
 
-  // TODO: Auth: Set or disable rate limit if authenticated user
-
   // Current route
   const path = req.baseUrl + req.path
-  const route = req.method + path.split("/").slice(0,4).join("/")
-
-  // Create new RateLimit if none exists for this route
-  if (!uniqueRateLimits[route]) {
-    uniqueRateLimits[route] = new RateLimit({
-      windowMs: 60 * 1000, // 1 minute window
-      delayMs: 0, // disable delaying - full speed until the max limit is reached
-      max: maxRequests, // start blocking after maxRequests
-      handler: function(req: express.Request, res: express.Response /*next*/) {
-        res.format({
-          json: function() {
-            res.status(500).json({
-              error: "Too many requests. Limits are 60 requests per minute."
-            })
-          }
-        })
-      }
-    })
+  const route =
```

Comment:
```suggestion
const rateLimitTier = req.locals.proLimit ? "PRO" : "BASIC"
  const route =
      rateLimitTier +
```

Reply:
The routes need to be distinct for PRO and BASIC tiers to create and use to both